### PR TITLE
RATEST-150R:New E2E Test For Registration Workflow

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/RATEST-214

### CODE REVIEW

**Comment by kdaud** [comment]
@sherrif10 rebuild the PR to eliminate other workflows to run along side the RefApp 2.x Search & Registration workflow.

**Comment by sherrif10** [comment]
> RefApp 2.x Search & Registration workflow.

We nolonger have this workflow. This work workflow will be replaced by the extraction of[ FindPatient](https://github.com/openmrs/openmrs-contrib-qaframework/pull/101) and [Registration](https://github.com/openmrs/openmrs-contrib-qaframework/pull/102) independent work which am currently working on

**Comment by kdaud** [comment]
Rebuild the PR to eliminate other workflows to run along side the PR changes!

**Comment by sherrif10** [comment]
Thats is still tricky. After rebuilding whole workflows are popped up

**Comment by kdaud** [comment]
@sherrif10 do you mind to trigger a commit to rebuild the PR and see the effect once again ?

**Comment by sherrif10** [comment]
Its a continual triggering of commit, but i realised sometimes , single commit rebuilds as expected.Thanks

**Comment by sherrif10** [comment]
Hello @kdaud , Any more recommendations

**Comment by kdaud** [comment]
Just as we discussed in our QA call we can merge this workflow test together with Search Patient so that we have a workflow that combines  both searching and Registering a patient in order to have it being tracked on our Dashboard. Some thing similar to RefApp 3.x Search & Registration

**Comment by sherrif10** [comment]
> Just as we discussed in our QA call we can merge this workflow test together with Search Patient so that we have a workflow that combines both searching and Registering a patient in order to have it being tracked on our Dashboard

**It seems you got it wrong. Let me be little clear here**
The reason why we are moving away from search and registration into same workflow is because, search functionalities are being captured/used in other workflows Fyi cucumber will a must throw NullpointerException, Am sure you have been following this implementation [here ](https://github.com/openmrs/openmrs-contrib-qaframework/pull/79/) if you looked so closely with that ticket, all the tests passed but because there are other functions that they are consuming, the configuration of ALLFIRREFOX would fail with exact same error.

**What we discussed yesterday o**n the call was to separate these tests into one single workflow, The idea behind this as said by **christine** was to have a independent workflow of two separate workflows and this will improve these workflows. Merging the two will take us back but it is also doable. However my work here is dependant on what we agreed on the call last nigh.Feel free to suggest incase am not clear please.Thanks.

**Comment by kdaud** [comment]
Christine's idea was to have Search & Registration workflow test being reflected on the QA Dashboard and my recommendation was to refactor FindPatientSteps and RegistrationSteps into one workflow to have the goal achieved.

**Comment by sherrif10** [comment]
> Christine's idea was to have Search & Registration workflow test being reflected on the QA Dashboard and my recommendation was to refactor FindPatientSteps and RegistrationSteps into one workflow to have the goal achieved.

Can we first move this conversation on qa-support slack channel to have better direction before i continue with this work.

**Comment by kdaud** [comment]
We may need her clear recommendation otherwise am not sure whether we will get her response via slack soon.

**Comment by sherrif10** [comment]
Lets wait for her , then i can proceed with this work , otherwise we might waste time implementing this differently.

**Comment by wanjikuchristine** [comment]
> > Just as we discussed in our QA call we can merge this workflow test together with Search Patient so that we have a workflow that combines both searching and Registering a patient in order to have it being tracked on our Dashboard
> 
> **It seems you got it wrong. Let me be little clear here**
> The reason why we are moving away from search and registration into same workflow is because, search functionalities are being captured/used in other workflows Fyi cucumber will a must throw NullpointerException, Am sure you have been following this implementation [here ](https://github.com/openmrs/openmrs-contrib-qaframework/pull/79/) if you looked so closely with that ticket, all the tests passed but because there are other functions that they are consuming, the configuration of ALLFIRREFOX would fail with exact same error.
> 
> **What we discussed yesterday o**n the call was to separate these tests into one single workflow, The idea behind this as said by **christine** was to have a independent workflow of two separate workflows and this will improve these workflows. Merging the two will take us back but it is also doable. However my work here is dependant on what we agreed on the call last nigh.Feel free to suggest incase am not clear please.Thanks.



> Lets wait for her , then i can proceed with this work , otherwise we might waste time implementing this differently but my direction would suffice and work better still.

Hey @sherrif10 and @kdaud ,  my recommendation was to keep the two functionalities separate due to the issues of duplication . However we should have the status labels each of the workflows present QA status project dashboard..

**Comment by sherrif10** [comment]
Thanks @wanjikuchristine  for clarification. cc @kdaud

**Comment by sherrif10** [comment]
Looking foward for approve from @kdaud  and go ahead merge from @dkayiwa .)

**Comment by sherrif10** [comment]
Hello @kdaud , Waiting for your final reviews then we get this merged and i switch remaining workflow

File: src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:24]: I expect @Before method some where since  `RegistrationSteps` is being detached from the orders dwelling in `qa.yml` to rather use orders for `registration.yml`

File: src/test/resources/features/refapp-2.x/03-createPatient/patientRegistration.feature
Comment (kdaud) [inline src/test/resources/features/refapp-2.x/03-createPatient/patientRegistration.feature:3]: Any reason for removing @selenium ?

**Comment by kdaud** [comment]
> Waiting for your final reviews then we get this merged and i switch remaining workflow

Once you request for review it may not be a good idea to wait the commit to be merged  then switch to other work. What if the PR takes a month or even beyond under coder review ?

I would ideally recommend once you request for review continue working on other tasks as you keep an eye on the reviews and  address them along the way while working on other tasks. git branching will achieve all for you.

I learnt the logic  from @dkayiwa !

File: src/test/resources/features/refapp-2.x/03-createPatient/patientRegistration.feature
Comment (sherrif10) [reply src/test/resources/features/refapp-2.x/03-createPatient/patientRegistration.feature:3]: Resolved thanks.

**Comment by dkayiwa** [comment]
> I learnt the logic from @dkayiwa !

Good student! 😊

File: src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java
Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:24]: Lets  not add before method because its not by default that we need before method every implementation. In before method, i would add a functionality of  initalizing loginPage  , However this is being capture behind the scenes

**Comment by sherrif10** [comment]
Hello @kdaud  @dkayiwa ,  Will be merging this because i will be inheriting some after methods in  SearchPatient workflow. Thanks. Incase we delay a bit.

File: src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java
Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:24]: > Lets not add before

Which logic will be responsible for the test login into the system ? And what will be the background of the test ?

**Comment by kdaud** [comment]
> i will be inheriting some after methods in SearchPatient workflow

Why making the two workflow tests  inter-depending on each other ?

File: src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java
Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:24]: By default goToHomePage will irrespective of adding a before method. Am also following the initial implementation

Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:24]: Now that we making the workflow to be independent so that we track its status on the Dashboard, we need to have its own logic that logs the test into the system and this can be achieved within @Before method.

> Am also following the initial implementation.

The initial implementation of the test was being governed by commands that are dwelling in `qa.yml,`  you need to make sure that  the workflow is governed by the commands from `registration.yml`

**Comment by kdaud** [comment]
@sherrif10 was it an oversight not to add the the ticket url ?

**Comment by sherrif10** [comment]
Resolved thanks

**Comment by sherrif10** [comment]
Hello @kdaud  ant more recommendations or reviews.

**Comment by kdaud** [comment]
Catch my comments [1](https://github.com/openmrs/openmrs-contrib-qaframework/pull/102#discussion_r702525554) and as well [2](https://github.com/openmrs/openmrs-contrib-qaframework/pull/102#issuecomment-912350707) !

File: .github/workflows/refapp-2x-registration.yml
Comment (sherrif10) [inline .github/workflows/refapp-2x-registration.yml:7]: Did you see this file? @kdaud

File: src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java
Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:24]: I already have an implementation written for [registration.yml](https://github.com/openmrs/openmrs-contrib-qaframework/pull/102/files)

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:24]: Another suggestion why we might not need before hook method is that we already have an implementation of login in[ here ](https://github.com/openmrs/openmrs-contrib-qaframework/pull/102/files#diff-68d8685a62f4a01878be56c191b10ca50a9a679b8d3047f0eedc2265d5370f79R32), so using[ initiateLogin()](https://github.com/openmrs/openmrs-contrib-qaframework/pull/101/files#diff-d7d118d304f2585ada64c935320d5efb7ca710d547dfb7aa45bc5b89165b073cR23) will throw duplicate exception.

**Comment by sherrif10** [comment]
Hello, @kdaud I addressed all your comments.

**Comment by kdaud** [comment]
@sherrif10 have you addressed the comment [here](https://github.com/openmrs/openmrs-contrib-qaframework/pull/102#issuecomment-912350707) ?

**Comment by sherrif10** [comment]
> Why making the two workflow tests inter-depending on each other?

What do you mean interdependent, these two tests are two different tests running independently,

**Comment by sherrif10** [comment]
Hello @kdaud  any more suggestions or recommendations.

**Comment by kdaud** [comment]
> What do you mean by interdependent, this test will run dependently.

Your comment [here,](https://github.com/openmrs/openmrs-contrib-qaframework/pull/102#issuecomment-912297243) it suggests that this workflow has some functionality that is  supposed to be exercised by another test!

**Comment by sherrif10** [comment]
> Your comment [here,](https://github.com/openmrs/openmrs-contrib-qaframework/pull/102#issuecomment-912297243) it suggests that this workflow has some functionality that is supposed to be exercised by another test!

As you christine clearly stated above , this is an dependent ticket and it only handles for registration workflows.

File: src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:23]: You can make the class variables private since they are only consumed within the class

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java]: Resolved thanks

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java]: Waiting for your final reviews @kdaud.Thanks

File: src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:30]: Should n't the the user story read as` Registered user rightly logs in` ?

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:30]: > Should n't the the user story read as` @Given("Registered user rightly logs in")` ?

Sure the user story is read as Given as seen, Any mistake with that

Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:30]: What do you say about `Registration user` vs `Registered user` ?

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:30]: > Registered user

Registered User sounds  good however it's different from registration user , ideally i think its the person who registers patient(registration clerk, registration nurse, registration doctor), not the already registered user(registered User. registered clerk registered nurse)

Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:30]: According to the test implementation, the user story is talking of  login in of a registered user within the system!

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:30]: > registered user

Its talking of registration user, the person  who registers a patient forexample nurse, doctor , these person will become a user after him/she logs into the system

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:30]: To be little more detailed: This functionality should be the one in before method. so it should bring out the aspect of registration logs in not user logs in

**Comment by sherrif10** [comment]
Any more recommendations @kdaud

File: src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java
Comment (kdaud) [inline src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:23]: @sherrif10 make use of inheritance and get rid of `ClinicianFacingPatientDashboardPage`

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:23]: > @sherrif10 make use of inheritance and get rid of `ClinicianFacingPatientDashboardPage`

ohh where would I use inheritance here, Do you mind suggest an example.

Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:23]: `RegistrationSteps` is a child class of `Steps` and  by default variables that are declared within the parent class can be used in a child class.

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:23]: > `RegistrationSteps` is a child class of `Steps` and by default variables that are declared within the parent class can be used in a child class.

is this really supposed to be implemented in this class workflow ? i think you are talking about something else

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:23]: @kdaud  any more reviews.

**Comment by sherrif10** [comment]
Merging this.Thanks @kdaud  for review

File: src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java
Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java:23]: Variables that are declared in a parent class can directly be used by a child class, so whenever a variable declared in a parent class you do not need to declare it in a child class to eliminate code redundancy !

@sherrif10 do you think you need again to declare this variable [here](https://github.com/openmrs/openmrs-contrib-qaframework/blob/d5f431cbc90972e1a5a6eca0b955f7edc26ee35d/src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java#L23) ?

Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java]: Resolved thanks.

**Comment by kdaud** [comment]
The user story in [PatientRegistration.feature](https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/src/test/resources/features/refapp-2.x/03-createPatient/patientRegistration.feature#L5)  that talks about logging into system  should be refactored to `Given Registered user rightly logs in`

**Comment by sherrif10** [comment]
> The user story in [PatientRegistration.feature](https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/src/test/resources/features/refapp-2.x/03-createPatient/patientRegistration.feature#L5) that talks about logging into system should be refactored to `Given Registered user rightly logs in`

It should be registration user not registered user please.did you see this comments above  [here](https://github.com/openmrs/openmrs-contrib-qaframework/pull/102#discussion_r703004304)

**Comment by kdaud** [comment]
> the person who registers a patient for example nurse, doctor, this person will become a user after he/she logs in to the system. we are assuming the system has no registered person data


> did you see this comments above [here](https://github.com/openmrs/openmrs-contrib-qaframework/pull/102#discussion_r703004304),

The user story is talking about a registered user of the system having a privilege of logging into the system once the entered credentials are right. The user can be admin. nurse, doctor or any user as long as is registered in the system.  Here is its [implementation](https://github.com/openmrs/openmrs-contrib-qaframework/blob/7ab2d2b514995693837fab5782cdd82d921955ec/src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java#L33)

After the user is logged into the system, can then register patient details as implemented [here](https://github.com/openmrs/openmrs-contrib-qaframework/blob/7ab2d2b514995693837fab5782cdd82d921955ec/src/test/java/org/openmrs/contrib/qaframework/automation/RegistrationSteps.java#L44)

**Comment by sherrif10** [comment]
@kdaud  do you see any difference btn that kind of implementation and https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/src/test/java/org/openmrs/contrib/qaframework/automation/Steps.java#L67  line

**Comment by sherrif10** [comment]
To me, i think a registered user means a person who is already registered in the databases and this simply logs in the system, however we are talking about registration nurse, the person who registers patients in the system, you can actually see the kind of code they used.

**Review by kdaud — APPROVED** [review]
Waiting for final reviews cc: @dkayiwa

**Comment by sherrif10** [comment]
Thanks @kdaud  for the reviews and suggestions.

**Comment by sherrif10** [comment]
Hey @dkayiwa , Merging this. @kdaud   have  approved.

**Comment by kdaud** [comment]
@sherrif10 before you merged did you take a look at my last comment [here](https://github.com/openmrs/openmrs-contrib-qaframework/pull/102#pullrequestreview-747557490) ?

**Comment by sherrif10** [comment]
I think @dkayiwa  would have merged it since you approved it.

**Comment by kdaud** [comment]
> I think @dkayiwa would have merged it since you approved it.

There has been times on PRs which personally have approved and later @dkayiwa in his final reviews points out critical areas in the PR of which I personally and PR author end up learning a thing!

I think  we need to use this grand opportunity to learn from our mentor's final reviews than just merging PRs without them (Ian and Daniel) taking a look at the PR.  I remember @dkayiwa recommending it to both of us unless I have just forgotten that he recommended to only @kdaud exempting @sherrif10 . @sherrif10  you can remind me if have forgotten !
 
@sherrif10 I would love to hear of what you think about merging PRs. Should we then be merging PRs as long as @sherrif10  or @kdaud  has approved the PR?

**Comment by kdaud** [comment]
@sherrif10 have you had a look the comment [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/102#issuecomment-914051052) ?

**Pull request merged**
